### PR TITLE
[MXNET-326] fix filter layout in DepthwiseConv2dBackwardFilterKernel

### PR DESCRIPTION
**Pull request by nihui**
## Description ##
the filter layout should be NCHW, while the exsiting code write as NHWC, which is wrong.
this change fix it.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/nn/depthwise_convolution_tf.cuh [hunk 1]
```diff
@@ -450,14 +453,17 @@ DepthwiseConv2dBackwardFilterKernel(const DepthwiseArgs args,
         const int input_offset_temp =
             (out_b * in_channel * in_height * in_width) +
             (in_c * in_height * in_width) + (in_row * in_width);
+        const int filter_backprop_temp =
+            (in_channel * filter_width * filter_height) +
+            (filter_width * f_h);
         CUDA_UNROLL for (int f_w = 0; f_w < filter_width; ++f_w) {
           const int in_col = in_col_start + f_w;
           const int addr_temp = filter_width * f_h;
```
Comment (xinyu-intel) [inline src/operator/nn/depthwise_convolution_tf.cuh]: It seems that variable `addr_temp` was declared but never referenced.

**Comment by xinyu-intel** [comment]
Thank you for this pr. You can also open the depth_wise_conv unit test in your pr to check if this change can be passed:)

**Comment by xinyu-intel** [comment]
Thank you:) It seems that this change can pass the unit test except MKLDNN version which will be fixed in #10365 .

**Comment by xinyu-intel** [comment]
@nihui I will apply my changes to your repo to check if this pr can solve both two problems:)

**Comment by pengzhao-intel** [comment]
Good job @nihui @xinyu-intel 

@zheng-da @marcoabreu please help take a review :) 
And we can close #8721 after this PR is merged.

File: tests/cpp/operator/mkldnn.cc [hunk 2]
```diff
@@ -77,4 +77,11 @@ TEST(MKLDNN_UTIL_FUNC, AlignMem) {
   LOG(INFO) << "Skipped for GCC " << __GNUC__ << "." << __GNUC_MINOR__;
 #endif
 }
+
+TEST(MKLDNN_UTIL_FUNC, MemFormat) {
+  // Check whether the number of format is correct.
+  CHECK_EQ(mkldnn_format_last, 56);
+  CHECK_EQ(mkldnn_nchw, 5);
+  CHECK_EQ(mkldnn_oihw, 12);
+}
```
Comment (zheng-da) [inline tests/cpp/operator/mkldnn.cc]: This is MKLDNN specific. Maybe it's better to keep it in the fix for the MKLDNN implementation.

File: src/operator/nn/mkldnn/mkldnn_base.cc [hunk 3]
```diff
@@ -239,6 +239,7 @@ mkldnn_memory_format_t GetDefaultFormat(const mkldnn::memory::desc &desc) {
       case mkldnn_gOihw8o:
       case mkldnn_Goihw8g:
       case mkldnn_gOihw16o:
+      case mkldnn_Goihw16g:
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_base.cc]: is this related to this PR?

**Comment by zheng-da** [comment]
I'm a little confused. Does this PR fix everything (including MKLDNN) or just fix the bug in the original implementation?

If this is to fix everything, I suppose we can close https://github.com/apache/incubator-mxnet/pull/10365?

**Comment by TaoLv** [comment]
@zheng-da Yes. #10365 can be closed. We merged these two PRs together, otherwise neither of them can pass ci.

**Comment by marcoabreu** [comment]
Thanks a lot for catching this and re-enabling the test, @nihui! Also thanks to the people from Intel to have these joint efforts!

**Comment by nihui** [comment]
I think it is needed to cherry-pick this commit to v1.2.0 branch, as depthwise convolution is widely used, otherwise a regression bug would be in the coming stable release. thanks! @marcoabreu
